### ISSUE TITLE
GPU training is slower than CPU

### ISSUE THREAD

**Issue by clavvis** [deeplearning4j/nd4j#2750]
I bought new GPU with GTX 1060 6GB. Set Nd4jBeckend as CUDA 8.0. I downloaded NVidia CUDA Toolkit 8.0 and patch 2 and both installed.
I set GPU memory limit 6GB.

I start training my neural network (I use dl4j 0.91) and I see it is slow. It is slower almost two times than CPU training.
GPU average load is 1-3 %. It is not training - it is only screen showing !!!
CPU load is 50-85%. In CPU training was 97-100%.
GPU memory usage is 2-3 GB.
My data is not big. I guess tents MB.

I use those commands:
logger.info("Nd4jBackend.BACKEND_PRIORITY_CPU: {}", Nd4jBackend.BACKEND_PRIORITY_CPU);
logger.info("Nd4jBackend.BACKEND_PRIORITY_GPU: {}", Nd4jBackend.BACKEND_PRIORITY_GPU);
//
logger.info("Nd4j.getBackend(): {}", Nd4j.getBackend().getClass().getSimpleName());
logger.info("getBackend.getPriority(): {}", Nd4j.getBackend().getPriority());
		
Console shows:
c.vhr.bac.BAC36 - Nd4jBackend.BACKEND_PRIORITY_CPU: 0
c.vhr.bac.BAC36 - Nd4jBackend.BACKEND_PRIORITY_GPU: 100
o.n.l.f.Nd4jBackend - Loaded [JCublasBackend] backend
o.r.Reflections - Reflections took 587 ms to scan 388 urls, producing 128519 keys and 140910 values 
o.n.n.NativeOpsHolder - Number of threads used for NativeOps: 32
o.r.Reflections - Reflections took 115 ms to scan 19 urls, producing 31 keys and 227 values 
o.n.l.a.o.e.DefaultOpExecutioner - Backend used: [CUDA]; OS: [Windows 10]
o.n.l.a.o.e.DefaultOpExecutioner - Cores: [8]; Memory: [3,6GB];
o.n.l.a.o.e.DefaultOpExecutioner - Blas vendor: [CUBLAS]
o.n.l.j.o.e.CudaExecutioner - Device name: [GeForce GTX 1060 6GB]; CC: [6.1]; Total/free memory: [6442450944]
c.vhr.bac.BAC36 - Nd4j.getBackend(): JCublasBackend
c.vhr.bac.BAC36 - getBackend.getPriority(): 100

I don't see error. Could you help me?

Thank you

**Comment by saudet** [comment]
GPUs are not always going to be faster, especially for small amounts of data. Try to increase at least the batch size to as large a value as you can.

**Comment by clavvis** [comment]
But GPU doesn't do any mathematic operations for training. It is not OK.

**Comment by raver119** [comment]
No miracles out there. GPU is used, but your data is probably tiny and your OS is probably windows. So data transfers and kernels execution price is never payed back.

Show your NN configuration and input gratis shape, and you'll get more detailed answer

**Comment by clavvis** [comment]
Here is NN config:
https://gist.github.com/clavvis/2f4b8a7b57258289387b051aba8b53fc

If I do training GPU load is 1-3 %. If I do NOT training GPU load is 1-3 %.  ???

What is input features shape?

**Comment by raver119** [comment]
1) Can i see NN config in java?
2) Input features shape is exactly what i've said: you send some INDArray as input features to your nn. So i wonder what's the shape of that features array.
3) If you use CUDA backend, you're guaranteed to use CUDA for all math in 0.9.1. On current master it's not simple atm, but for 0.9.1 that's guaranteed.

**Comment by clavvis** [comment]
ad 1. I have something like Arbitr and set NN config using ParametrSpace class. Do you want it?
   CPU training works OK.

ad 2. I use DataSetIterators for MLN.pretrain and MLN.fit
My data are time series and are normalized to range 0-1.
It is 18667 rows and 325 columns. How send it?

ad 3. I use dl4j from jars (maven repozitories). All is 0.9.1.

**Comment by raver119** [comment]
1. yes please. anything that's human readable should work for me.
2. 18667 rows - that's total number of examples you have? and what's your batch size?
3. ok.

**Comment by clavvis** [comment]
1. Here is: https://gist.github.com/clavvis/c69d3fd99497c592f522a6fcaaccb217
2. Data has 10 batches. Last version was "only" 18660 rows per 1866 row in batch.

**Comment by raver119** [comment]
Hm. Thats suspicious. 

Can you please attach PerformanceListener to your model, and provide me with its output?

**Comment by clavvis** [comment]
I have my stats. https://gist.github.com/clavvis/7cc2ad1e6a43198249907f1eb63bd46a
Traininig (pretrain include) start at 20:22:17. I stopped it at 07:29:36 next day. It was too slow.
Every epoch is 10 batches. Maybe it could be enough information.

**Comment by treo** [comment]
You were asked for the PerformanceListener Output, as it contains more information than just how long the overall thing takes.

In cases like yours it isn't unusual that just creating the next DataSet object takes long enough that the GPU sits around twiddling its proverbial thumbs, while it waits for more data to arrive. 

The PerformanceListener output will tell you how long it takes to do exactly that.

**Comment by raver119** [comment]
Exactly. I’ve asked for PerformanceListener output because of its ability to show ETL and training time separately.

**Comment by clavvis** [comment]
I can do it. But standard proces take about 15 hours. Can I use smaller epochs count?

**Comment by raver119** [comment]
i just need to see output for 10-20-30 iterations. No need for full training cycle :)

**Comment by clavvis** [comment]
I use custom MLN = child of MLN https://gist.github.com/clavvis/b1a1a7e02e569280fb2775d9cd1edba5
Because standard MLN (0.9.1) do in fit(DataSetIterator) in one epoch pretrain and training too.
I hope it is not source og GPU problem.

**Comment by clavvis** [comment]
Here are results for one training MLN: https://gist.github.com/clavvis/263a013ed478e3df813458d2c0a897ef

**Comment by clavvis** [comment]
Here are results for 3 MLN training at one time: https://gist.github.com/clavvis/edefd5ad533af7e6b543fb2df1309769

**Comment by raver119** [comment]
You have very big time variance there. It shouldn’t be like that.

**Comment by clavvis** [comment]
Here is my DataProvider and DataSetItrator. https://gist.github.com/clavvis/dbf8a8b3d50ee5785dcf8f7721ad803c

**Comment by AlexDBlack** [comment]
> You have very big time variance there. It shouldn’t be like that.

Looking at the results for the single and 3-network execution: These spikes look at first glance a *lot* like garbage collector (GC) pauses due to huge number of objects in JVM.

@clavvis Can you run this with a profiler attached? Yourkit or visualvm. I'd like to see 2 things:
(a) Total number of objects in the JVM during execution
(b) The GC activity/pauses

And make sure you are using workspaces *and* GC configuration as per here: https://deeplearning4j.org/workspaces

**Comment by clavvis** [comment]
First I did some fast tests. Settings: 50 epochs pretrain  and 100 epoch training. 10 batches.

0. duration 8.29 minute - big iterations duration variances

1. add command:  Nd4j.getMemoryManager().setAutoGcWindow(5000);
   duration 2.51 minute - big it. dur. variances

2. GC command changed to: Nd4j.getMemoryManager().togglePeriodicGc(false);
  duration 2.00 minute - smaller it. dur. variances, but still distinctive

3. 1 batch
  duration 1.51 minute - small iterations (=epoch now)  duration variances
  iteration duration  is 680-714 ms (38x) and other are 563, 935 and 779 ms

4. 2 batches
  duration 1.55 minute - iterations duration is 51, 620, 53, 618 and further similar

**Comment by raver119** [comment]
1) Please always post PerformanceListener results. We're able to interpret them :)

2) Had you enabled workspaces?

**Comment by clavvis** [comment]
Training and inference Workspacemode is set to SEPARATE. P.lister results I added.

**Comment by raver119** [comment]
Get any java profiler tool (ie JVisualVm or YourKit profiler) and make screenshot of GC tab please.

**Comment by raver119** [comment]
p.s. it definitely looks like GC problem.

**Comment by clavvis** [comment]
With 1 batch or with another batches count? What should be GC settings?

**Comment by AlexDBlack** [comment]
@clavvis for the GC window output, let it run for a bunch of batches... doing training for a few minutes should be enough. As for GC setting - doesn't matter too much - but say run with every 5 seconds (default, or every 10 seconds would also be ok), just mention what settings you used.

**Comment by clavvis** [comment]
It is for 10 batches and GC every 5 sec.
![jvisualvm_01](https://user-images.githubusercontent.com/31794846/37648105-46020938-2c2e-11e8-96d2-18c568b8ce3c.PNG)

**Comment by AlexDBlack** [comment]
OK, so that's a bit less useful than I had hoped... still, it confirms that there's significant GC activity.
Assuming I'm reading this correctly - note the 190 total collections, 34.9 sec... 25 old gen collections, 31.1 sec
So it's mostly old gen. And it suggests there's a lot of long-lived objects - which is known to cause long GC pauses.

So in my mind, the only remaining questions here are:
(a) how many objects?
(b) what are the objects?

To answer both questions, check visualvm: Applications -> Heap dump -> Objects (drop-down)

**Comment by raver119** [comment]
I have way more simple question. Is that our code used for everything? I.e. Arbiter + DataVec + DL4j, or there's someone else's code as well in the middle?

**Comment by clavvis** [comment]
![jvisualvm_02](https://user-images.githubusercontent.com/31794846/37721417-ab038676-2d29-11e8-943f-2a2fb2581402.PNG)

First named items are my data.

**Comment by clavvis** [comment]
Yes

**Comment by clavvis** [comment]
For "deleting" all useless objects should I set null all objects, or is enough set null only "super" objects (arrays, lists)?

**Comment by raver119** [comment]
It's not that easy. You have no problems "deleting" useless objects, GC does that for you, and THAT is your problem. There are ways to get around that obviously, but that has nothing to do with nullified objects etc

**Comment by clavvis** [comment]
I have super object useful, but some inside object useless. I set some null; Now it look better.
![jvisualvm_03](https://user-images.githubusercontent.com/31794846/37728503-b6172ba6-2d3a-11e8-86e2-f2f3a225539f.PNG)

**Comment by AlexDBlack** [comment]
Yeah, 6.5 million Double objects - that's definitely certainly the cause here.
That'll cause bad GC performance and long pauses it *any* Java code - though ND4J will be a more prone to it due to (much decreased with workspaces) reliance on GC for off-heap memory.

Millions of Double objects points to bad design, IMO - both in terms of memory and performance.
I'm assuming it's coming from List<Double> and similar? If so, I'd suggest using double[], INDArray or something like Eclipse collections DoubleArrayList - https://github.com/eclipse/eclipse-collections/blob/master/docs/guide.md#primitive-lists

Beyond that, tuning GC frequency and using workspaces will help, but fixing the underlying design issue will be a much better solution.

As the GC pressure isn't due to our code - I'm closing this issue.

**Comment by clavvis** [comment]
Thanks for help.
